### ISSUE TITLE
capture output programmatically

### ISSUE THREAD

**Issue by ethanabrooks** [xd009642/tarpaulin#357]
Hi,
I am writing an autograder that uses code coverage as part of students' grades. I am wondering if there is a way to use this tool from within a rust program (the autograder).

Related to this, I notice that even when I specify the `--out` flag, the actual output to stdout does not change. Instead tarpaulin generates a new file (e.g. coberta.xml). Is there a way to instead output the contents of that file to stdout so that I might capture it `std::process::Command`?

**Comment by xd009642** [comment]
I don't see any reason why `std::process::Command` wouldn't work 👍

**Comment by ethanabrooks** [comment]
The trouble is that the output is not in a standard format:
```
❯ cargo tarpaulin --out Json

[INFO tarpaulin] Running Tarpaulin
[INFO tarpaulin] Building project
[INFO tarpaulin] Launching test
[INFO tarpaulin] running /home/ethan/autograder/autograder/submission/target/debug/deps/submission-cc8c238ef2ff13f5

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

[INFO tarpaulin] Launching test
[INFO tarpaulin] running /home/ethan/autograder/autograder/submission/target/debug/deps/unit-d87c04ba982f32fd

running 8 tests
...F....
failures:

---- it_doesnt_work stdout ----
Error: "two plus two does not equal three"
thread 'it_doesnt_work' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`: the test returned a termination value with a non-zero status code (1) which indicates a failure', <::std::macros::panic macros>:5:6
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    it_doesnt_work

test result: FAILED. 7 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

[INFO tarpaulin] Coverage Results:
|| Tested/Total Lines:
|| src/lib.rs: 5/6 +0%
|| tests/unit.rs: 22/22 +0%
|| 
96.43% coverage, 27/28 lines covered, +0% change in coverage
Error: "Output format is currently not supported!"
```
Converting that output to a string and parsing it would be asking for trouble, I think.

**Comment by xd009642** [comment]
Ah right, well JSON output isn't supported yet, there is XML. Could add a print format option to solve this but then have to think about where my other logging will go so it doesn't mess it up

**Comment by ethanabrooks** [comment]
Looking at the source, it looks like several of your functions have a `Result<(), RunError>` return type. I would be willing to put together a pull request where the return type is instead `Result<String, RunError>` where the `String` corresponds to the output that finally gets written to disk. Would you be amenable to this?

**Comment by ethanabrooks** [comment]
It looks like your `export` method (e.g. https://github.com/xd009642/tarpaulin/blob/master/src/report/cobertura.rs#L122) both builds a byte string and writes to disk. Seems like we would just need to separate these functionalities into two separate methods.

**Comment by xd009642** [comment]
Yeah I'm open for such a PR as long as it doesn't break any existing features. Not all the exporters will write to disk as some like coveralls are just sending to a HTTP endpoint

**Comment by xd009642** [comment]
Also, I'm kinda curious about your use-case (just in terms of where tarpaulin is finding itself used), like what sort of students will be using this, with what other tools, and what sort of platform?

**Comment by ethanabrooks** [comment]
This is an undergraduate level class focused on the theoretical foundations
of programming languages. The first half focused on the Lambda Calculus and
the second will begin to introduce features found in imperative languages.
We will do some assignments on Rust’s lifetime and borrowing concepts and
also an assignment where students use Rust for concurrency.

**Comment by ethanabrooks** [comment]
Ok I just opened this PR: https://github.com/xd009642/tarpaulin/pull/362

I am a little new to Rust and one thing that is currently throwing me a little is the fact that your `toml` file is named `tarpaulin.toml` instead of `Cargo.toml`. This is throwing off the `Cargo.toml` file in my project, because when I try to reference my forked version of tarpaulin by git url or or local path, it claims that it can't find the project:
```
error: no matching package named `tarpaulin` found
location searched: https://github.com/ethanabrooks/tarpaulin.git
```
Is it possible either to rename to `Cargo.toml` or to point my package to your `tarpaulin.toml` file?

**Comment by xd009642** [comment]
Is this issue adequately addressed after #362 was merged and a new version is released

**Comment by xd009642** [comment]
I'm going to close this issue, feel free to open another for anything else you need :+1:
